Source/MockCapture.cs

```diff
@@ -0,0 +1,131 @@
+//INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+//WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+//NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+//OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+//SUCH DAMAGE.
+
+//[This is the BSD license, see
+// http://www.opensource.org/licenses/bsd-license.php]
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace Moq
+{
+    /// <summary>
+    /// Captures a mock parameter.
+    /// </summary>
+    /// <example>
+    /// Arrange code:
+    /// <code>
+    /// var capture = new MockCapture{string}();
+    /// mock.Setup(x => x.DoSomething(capture.AsParameter()));
+    /// </code>
+    /// Assert code:
+    /// <code>
+    /// Assert.Equal("Hello!", capture.Last);
+    /// </code>
+    /// </example>
+    public class MockCapture<T>
+    {
+        private readonly List<T> _capturedValues = new List<T>();
+        private Action<T> _callback;
+        private Func<T, T> _copier;
+
+        /// <summary>
+        /// Gets all the captured parameter values.
+        /// </summary>
+        public IEnumerable<T> All { get { return _capturedValues; } }
+
+        /// <summary>
+        /// Gets the last captured parameter value.
+        /// </summary>
+        public T Last { get { return _capturedValues.Last(); } }
+        
+        /// <summary>
+        /// Gets the single captured parameter value.
+        /// </summary>
+        public T Single { get { return _capturedValues.Single(); } }
+
+        /// <summary>
+        /// Gets the parameter to use in the setup expression.
+        /// </summary>
+        /// <returns>The setup expression parameter</returns>
+        public T CaptureAny()
+        {
+            return Match.Create(value => CaptureParameter(value), () => It.IsAny<T>());
+        }
+
+        /// <summary>
+        /// Gets the parameter to use in the setup expression.
+        /// </summary>
+        /// <returns>The setup expression parameter</returns>
+        public T Capture(Expression<Func<T, bool>> match)
+        {
+            var matchDelegate = match.Compile();
+            return Match.Create(value => matchDelegate.Invoke(value) && CaptureParameter(value), () => It.Is<T>(match));
+        }
+
+        /// <summary>
+        /// Copy captured parameters with the specified delegate.
+        /// </summary>
+        /// <param name="copier">The delegate used to clone parameters</param>
+        public void CopyOnCapture(Func<T, T> copier)
+        {
+            _copier = copier;
+        }
+
+        /// <summary>
+        /// Provide an action to run on captured parameters.
+        /// </summary>
+        /// <paramref name="callback">an action to run on the captured parameters</paramref>
+        public void RunOnCapture(Action<T> callback)
+        {
+            _callback = callback;
+        }
+
+        private bool CaptureParameter(T item)
```

Comment:
If this method were, say, `protected virtual OnCapture(T item)`, a derived class in your test project may provide additional behavior on the parameter as needed, including creating a copier and/or the run on capture stuff.